codyze-compliance/src/main/kotlin/de/fraunhofer/aisec/cpg/codyze/compliance/SecurityGoal.kt

```diff
@@ -0,0 +1,112 @@
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.codyze.compliance
+
+import com.charleskorn.kaml.Yaml
+import de.fraunhofer.aisec.cpg.graph.Component
+import de.fraunhofer.aisec.cpg.graph.Name
+import de.fraunhofer.aisec.cpg.graph.OverlayNode
+import java.io.File
+import java.nio.file.Path
+import kotlin.io.path.Path
+import kotlin.io.path.extension
+import kotlin.io.path.readText
+import kotlin.io.path.walk
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+
+@Serializable
+data class SecurityGoal(
+    @Serializable(with = NameSerializer::class) override var name: Name,
+    val description: String,
+    val components: List<@Serializable(with = ComponentSerializer::class) Component?> = listOf(),
+    val assumptions: List<String> = listOf(),
+    val restrictions: List<String> = listOf(),
+    val objectives: List<SecurityObjective>,
+) : OverlayNode()
+
+@Serializable
+class SecurityObjective(
+    @Serializable(with = NameSerializer::class) override var name: Name,
+    val description: String,
+    val statements: List<String>,
+    val components: List<@Serializable(with = ComponentSerializer::class) Component?> = listOf(),
+    val assumptions: List<String> = listOf(),
+    val restrictions: List<String> = listOf(),
+) : OverlayNode()
+
+/** A custom serializer for the [Name] class. */
+class NameSerializer : KSerializer<Name> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor(Name::class.qualifiedName!!, PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Name) {
+        encoder.encodeString(value.localName)
+    }
+
+    override fun deserialize(decoder: Decoder): Name {
+        return Name(decoder.decodeString())
+    }
+}
+
+/** A custom serializer for the [Component] class. */
+class ComponentSerializer : KSerializer<Component?> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor(Component::class.qualifiedName!!, PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Component?) {
+        if (value != null) {
+            encoder.encodeString(value.name.toString())
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): Component? {
+        // TODO: find component by name somehow
+        return null
+    }
+}
+
+/** Load all security goals from a directory. */
+fun loadSecurityGoals(directory: Path): List<SecurityGoal> {
```

Comment:
Please support streams, too. See https://github.com/Fraunhofer-AISEC/cpg/issues/1935

Reply:
Only possible for the "single" version (below)